pennylane/tape/tape.py

```diff
@@ -872,6 +879,267 @@ def set_parameters(self, params, trainable_only=True):
+
+            elif ret_type == qml.operation.Probability:
+
+                wires_num_set = {len(meas.wires) for meas in mps}
+                same_num_wires = len(wires_num_set) == 1
+                if same_num_wires:
+                    # All probability measurements have the same number of
+                    # wires, gather the length from the first one
+
+                    len_wires = len(mps[0].wires)
+                    dim = QuantumTape._get_num_basis_states(len_wires, device)
+                    shot_copies_sum = sum(s.copies for s in shot_vector)
+                    shape = (shot_copies_sum, len(mps), dim)
+
+                else:
+                    # There are a varying number of wires that the probability
+                    # measurement processes act on
+                    # TODO: revisit when issues with this case are resolved
+                    raise UnsupportedTapeOperationError(
+                        "Getting the output shape of a tape with multiple probability measurements "
+                        "along with a device that defines a shot vector is not supported."
+                    )
+
+            elif ret_type == qml.operation.Sample:
+                shape = []
+                for shot_val in device.shot_vector:
+                    for _ in range(shot_val.copies):
+                        shots = shot_val.shots
+                        if shots != 1:
+                            shape.append(tuple([shots, len(mps)]))
+                        else:
+                            shape.append((len(mps),))
+
+        return shape
+
+    def get_output_shape(self, device):
+        """Produces the output shape of the tape by inspecting its measurements
+        and the device used for execution.
+
+        Note: as the output shape may be dependent on the device used for
+        execution, tapes do not store the computed shape.
+
+        Args:
+            device (~.Device): the device that will be used for the tape execution
+
+        Raises:
+            UnsupportedTapeOperationError: raised for unsupported cases for
+                example when the tape contains heterogeneous measurements
+
+        Returns:
+            Union[tuple[int], list[tuple[int]]]: the output shape(s) of the
+            tape result
+        """
+        output_shape = tuple()
+
+        if len(self._measurements) == 1:
+            output_shape = self._single_measurement_shape(self._measurements[0], device)
+        else:
+            num_measurements = len(set(meas.return_type for meas in self._measurements))
+            if num_measurements == 1:
+                output_shape = self._multi_homogenous_measurement_shape(self._measurements, device)
+            else:
+                raise UnsupportedTapeOperationError(
+                    "Getting the output shape of a tape that contains multiple types of measurements is unsupported."
+                )
+        return output_shape
+
+    def get_output_domain(self):
```

Comment:
I am very tempted to suggest we go with `tape.shape` and `tape.dtype` to match the other frameworks :) It feels like a very nice analogy for representing the output shape and dtype of an object

Reply:
Going with `tape.shape` and for the other chose `tape.numeric_type` (though could be renamed). For the latter, we might not have enough information about the:
* Interface
* Device

to determine the exact `dtype`. So the method would return one out of `int`, `float` or `complex` and then e.g., the QNode could define it's own `dtype` method using `tape.numeric_type`, device (for precisions) and the interface (for the exact object, e.g., `autoray` allows conversions between framework-specific dtypes).